src/box/lua/connpool.lua

```diff
@@ -73,77 +76,95 @@ local function connect(instance_name, opts)
     -- established or an error occurs (including a timeout error).
     if opts.wait_connected ~= false and conn:wait_connected() == false then
         local msg = 'Unable to connect to instance %q: %s'
-        error(msg:format(instance_name, conn.error), 0)
+        return nil, msg:format(instance_name, conn.error)
     end
     return conn
 end
 
--- Run connecting to the instance, which if it is not already
--- connected or in process of connecting.
-local function connect_in_background(instance_name)
-    pcall(connect, instance_name, {
-        wait_connected = false,
-        connect_timeout = WATCHER_TIMEOUT
+--- This method connects to the specified instances and returns
+--- the set of successfully connected ones.
+---
+--- Specifying a callback accepting an instance name and returning
+--- a boolean value in `opts.any` changes the behavior to connect
+--- to multiple instances until there is at least one satisfying
+--- the condition or the timeout occurs.
+function pool_methods.connect_to_multiple(self, instances, opts)
+    checks('table', 'table', {
+        any = '?function',
+        timeout = '?number',
     })
-end
 
-local function is_candidate_connected(candidate)
-    local conn = connections[candidate]
-    return conn and conn.state == 'active' and conn:mode() ~= nil
-end
-
--- Checks whether the candidate has responded with success or
--- with an error.
-local function is_candidate_checked(candidate)
-    local conn = connections[candidate]
+    if opts == nil then opts = {} end
+    if next(instances) == nil then return {} end
+    assert(opts.any == nil or type(opts.any) == 'function')
 
-    return not conn or
-           is_candidate_connected(candidate) or
-           conn.state == 'error' or
-           conn.state == 'closed'
-end
+    -- Checks whether the candidate is connected and active.
+    local function is_instance_connected(instance_name)
+        local conn = self._connections[instance_name]
+        return conn and conn.state == 'active' and conn:mode() ~= nil
+    end
 
--- This method connects to the specified instances and returns
--- the set of successfully connected ones.
---
--- opts.any could be used to specify a condition /w
-local function connect_to_candidates(candidates, opts)
-    if opts == nil then opts = {} end
-    if next(candidates) == nil then return {} end
+    -- Checks whether the candidate has responded with success or
+    -- with an error.
+    local function is_instance_checked(instance_name)
+        local conn = self._connections[instance_name]
+        return is_instance_connected(instance_name) or
+               conn.state == 'error' or conn.state == 'closed'
+    end
```

Comment:
I'm a bit worrying about lack of the `conn == nil` condition here.

I understand that we can't have `conn == nil` after `self:connect()` even after a yield in the current implementation.

However, how about updating the configuration that we want to implement later? Or autoclosing of unused connections?

I think that the functions performs a very small job and we shouldn't think about outside code (whether the `nil` is possible) while looking on it. I would handle this case too.